File: elixir/word-count/word_count.exs
defmodule Words do
  @doc """
  Count the number of words in the sentence.

  Words are compared case-insensitively.
  """
  @spec count(String.t()) :: map
  def count(sentence) do
    sentence
    |> String.downcase
    |> tokenise
    |> Enum.reduce(%{}, fn ([word], words) -> 
      Map.update words, word, 1, &(&1 + 1)
    end)
  end

  @spec tokenise(String.t()) :: [String.t()]
  defp tokenise(sentence) do
    Regex.scan ~r/(*UTF)[\p{L}0-9-]+/i, sentence
  end
end

File: typescript/word-count/word-count.ts

export default class Words {
    public count(words: string): Map<string, number> | undefined {
        const tokens = words.toLowerCase()
                            .match(/\S+/g) // Tokenise by whitespace
        // Check if result returned by match() is not null
        if (tokens) {
            return tokens.reduce((accumulator, item) => {
                return accumulator.set(item, accumulator.get(item) + 1 || 1)
            }, new Map())
        }
        return undefined
    }
}